src/main/telemetry/ibus.c

```diff
@@ -0,0 +1,387 @@
+    transmitIbusPacket(sendBuffer, sizeof sendBuffer);
+}
+
+static bool isCommand(ibusCommand_e expected, uint8_t ibusPacket[static IBUS_MIN_LEN]) {
+    return (ibusPacket[1] & 0xF0) == expected;
+}
+
+static ibusAddress_t getAddress(uint8_t ibusPacket[static IBUS_MIN_LEN]) {
+    return (ibusPacket[1] & 0x0F);
+}
+
+static void dispatchMeasurementRequest(ibusAddress_t address) {
+    switch (SENSOR_ADDRESS_TYPE_LOOKUP[address - ibusBaseAddress]) {
+    case IBUS_SENSOR_TYPE_EXTERNAL_VOLTAGE:
+        {
+            uint16_t value = vbat * 10;
+            if (ibusTelemetryConfig()->report_cell_voltage) {
+                value /= batteryCellCount;
+            }
+            sendIbusMeasurement(address, value);
+        }
+        break;
+
+    case IBUS_SENSOR_TYPE_TEMPERATURE:
+        {
+            #ifdef BARO
+                float temperature = (baroTemperature + 50) / 100.f;
+            #else
+                float temperature = telemTemperature1 / 10.f;
+            #endif
+            sendIbusMeasurement(address, (uint16_t) ((temperature + 40)*10));
+        }
+        break;
+
+    case IBUS_SENSOR_TYPE_RPM:
+        sendIbusMeasurement(address, (uint16_t) rcCommand[THROTTLE]);
+        break;
+    }
+}
+
+static void respondToIbusRequest(uint8_t ibusPacket[static IBUS_RX_BUF_LEN]) {
+    ibusAddress_t returnAddress = getAddress(ibusPacket);
+    
+    //autodetect first received address as our base address
+    if ((INVALID_IBUS_ADDRESS == ibusBaseAddress) && 
+        (INVALID_IBUS_ADDRESS != returnAddress))
+    {
+        ibusBaseAddress = returnAddress;
+    }
+
+    if ((returnAddress >= ibusBaseAddress) &&
+        (ibusAddress_t)(returnAddress - ibusBaseAddress) < sizeof SENSOR_ADDRESS_TYPE_LOOKUP) {
+        if (isCommand(IBUS_COMMAND_DISCOVER_SENSOR, ibusPacket)) {
+            sendIbusCommand(returnAddress);
+        } else if (isCommand(IBUS_COMMAND_SENSOR_TYPE, ibusPacket)) {
+            sendIbusSensorType(returnAddress);
+        } else if (isCommand(IBUS_COMMAND_MEASUREMENT, ibusPacket)) {
+            dispatchMeasurementRequest(returnAddress);
+        }
+    }
+}
+
+static void pushOntoTail(uint8_t buffer[IBUS_MIN_LEN], size_t bufferLength, uint8_t value) {
+    for (size_t i = 0; i < bufferLength - 1; i++) {
+        buffer[i] = buffer[i + 1];
+    }
+    ibusReceiveBuffer[bufferLength - 1] = value;
+}
+
+void initIbusTelemetry(void) {
+    ibusSerialPortConfig = findSerialPortConfig(FUNCTION_TELEMETRY_IBUS);
+    ibusPortSharing = determinePortSharing(ibusSerialPortConfig, FUNCTION_TELEMETRY_IBUS);
+    ibusBaseAddress = INVALID_IBUS_ADDRESS;
+}
+
+void handleIbusTelemetry(void) {
+    if (!ibusTelemetryEnabled) {
+        return;
+    }
+
+    while (serialRxBytesWaiting(ibusSerialPort) > 0) {
+        uint8_t c = serialRead(ibusSerialPort);
+        pushOntoTail(ibusReceiveBuffer, IBUS_RX_BUF_LEN, c);
```

Comment:
This code is quite fragile (it will be hard to support longer request packet) and is a bit suboptimal (each packet is tested four times / processed data are scanned again)
Interpacket gap delay be better (SBUS and others use that) ..